Return super in ActionController::Parameters.const_missing

shuhei opened the pull request:
The current implementation of `ActionController::Parameters.const_missing` returns `ActionController::Parameters.always_permitted_parameters` even if its `super` returns a constant without raising error. This prevents its subclass in a autoloading module/class from taking advantage of
autoloading constants.

``` rb
class SomeController < ActionController::Base
  def create
    SomeParameters.new(some_params).do_something
  end

  class SomeParameters < ActionController::Parameters
    def do_something
      DefinedSomewhere.do_something
    end
  end
end
```

In the code above, `DefinedSomewhere` is to be autoloaded with `SomeController.const_missing` but `ActionController::Parameters.const_missing` returns `always_permitted_parameters` instead of the autoloaded constant.

This pull request fixes the issue respecting `const_missing`'s `super`.

Commit:
Return super in ActionController::Parameters.const_missing

The current implementation of ActionController::Parameters.const_missing
returns `ActionController::Parameters.always_permitted_parameters` even
if its `super` returns a constant without raising error. This prevents its
subclass in a autoloading module/class from taking advantage of
autoloading constants.

    class SomeParameters < ActionController::Parameters
      def do_something
        DefinedSomewhere.do_something
      end
    end

In the code above, `DefinedSomewhere` is to be autoloaded with
`Module.const_missing` but `ActionController::Parameters.const_missing`
returns `always_permitted_parameters` instead of the autoloaded
constant.

This pull request fixes the issue respecting `const_missing`'s `super`.

fxn commented:
These kind of callbacks should fallback to `super`, but the reasoning doesn't sound correct to me.

The invocation of `const_missing` is a regular method call. When (the inherited) `SomeParameters.const_missing` is called, `SomeController.const_missing` plays no role, only the ancestors of `SomeParameters` do.

I have not run the test, but if you replace `assert` with `assert_equal` it should not pass.

shuhei commented:
@fxn Thanks for pointing out! I fixed the `assert` to `assert_equal` but it still passes.

Your explanation is right on plain Ruby but not on Rails. `active_support/dependencies` modifies `Module.const_missing` to look up `const_missing` of the module's parent. So, the chain would look like the following.

1. `SomeParameters.const_missing` (not defined)
2. `ActionController::Parameters.const_missing`
3. `Module.const_missing`
4. `SomeController.const_missing`
5. ... (not related to this PR)
6. autoloading from another file

So, the autoloaded constant should be returned by 3 and 2 should pass it through.

shuhei commented:
@fxn Ah, no. You are right. The 3 (Module.const_missing) autoloads the constant from another file.

fxn commented:
Yes, yes, I know Rails defines a global `const_missing`.

But that doesn't matter, `SomeController` is not an ancestor of `SomeParameter`, therefore `super` is going to hit the one in `Module`, not the one in `SomeController`.

shuhei commented:
@fxn OK, I'll fix the commit message.

shuhei commented:
Fixed the commit message and removed the test case. Now it doesn't seem that we need a test for this trivial fix.

I learned a lot about [constant resolution](http://guides.rubyonrails.org/autoloading_and_reloading_constants.html) in Rails through this PR. Thanks!

shuhei commented:
The test was passing because `Module.const_missing` actually [looks up its parent module](https://github.com/rails/rails/blob/v4.2.1/activesupport/lib/active_support/dependencies.rb#L526) after trying autoloading from files. But it's not straightforward example though for this fix.

fxn commented:
Ah... I see. Yes, it is done that way to "walk up the nesting" (ideally, with assumptions) when autoloading constants in Rails.

rafaelfranca commented:
Does we need to test it?

fxn commented:
I think we should, how could we test it without leaking stuff into the runtime of other tests?

shuhei commented:
It seems that [testing autoloading](https://github.com/rails/rails/blob/master/activesupport/test/dependencies_test.rb) looks too tedious for this fix. So I've added a test case to check whether the constant resolution algorithm works correctly.

fxn commented:
FWIW, I am not sure that `const_missing` call belongs to the public interface. The contract is that the parent looks up the constant, but it could be the case that it does so calling some other internal method.

What do you think about checking that `super` gets called without relying on AS::Dependencies? Like, injecting an ancestor or something like that.

shuhei commented:
Yes, I agree. The point is that `super` gets called regardless of AS::Dependencies. So I fixed the test to stub `const_missing` of `ActionController::Parameters`' superclass (AS::HashWithIndifferentAccess in the current implementation). Now it's become much simpler!

fxn commented:
Great test! Thanks for working on this patch :).

shuhei commented:
Thanks for guiding me to the right direction!
